File: scripts/mutation/listenerdiscovery.py
"""M-DISCOVERY: is socket OWNERSHIP what selects the process, or just whichever node is handy?

WRITE/EDIT TOOL ONLY -- heredocs in this environment eat backslashes.

`discoverListener` must answer "which process owns the socket listening on this port" from kernel
evidence, and nothing else. The named ways to get it wrong -- process name, image path, command
line, cwd, age, whichever PID makes a test pass -- would each select a plausible process on a
machine running many node processes that share one `node_modules`, and would be stably wrong.

Two mutants, and the second is the one that matters, because a name-based selector is what a
hurried implementation reaches for and it passes any test that only asserts "some listener was
found".

Expected cardinalities, written before the run:

  M-DISCOVERY-FIRSTROW take the first owner rather than requiring exactly one
                       -> 2 red: the two-distinct-owner control and the unparseable-row control.
                          This is the Windows defect review found, and the rule is now shared, so
                          one mutant binds both platforms.

  M-DISCOVERY-RACE     drop the two-observation agreement CHECK from the dispatcher
                       -> MISSED is the honest expectation. Reaching it needs a socket handoff
                          between two reads, which is not producible from a deterministic test. The
                          guard rests on reasoning; recorded, not rounded up.

  M-DISCOVERY-AGREE    make the agreement RULE itself always say yes
                       -> 2 red. The rule is now an exported pure function, so unlike the
                          dispatcher wiring above it can be bound by controls: the pid-reuse case
                          and the missing-observation case both fail.

  M-DISCOVERY-SELF     select the current process instead of the socket owner
                       -> 4 red, measured; I predicted 2 and the number is corrected rather than
                          the prediction re-read. The same-process control cannot catch it -- the
                          test owns the listener, so `process.pid` is the right answer there. The
                          killers are the controls asserting discovery goes NULL when nothing is
                          listening, plus the ones that reach a verdict through real discovery.

A FIRST ATTEMPT AT THIS MUTANT WAS BROKEN, and is recorded rather than quietly replaced. It put the
fallback on the FIRST observation only (`discover(port) ?? {...}`), so the second observation still
returned null for a closed port and the dispatcher refused anyway. It came back MISSED -- which I
had predicted, for a reason that turned out to be wrong. A MISSED predicted for the wrong reason is
a broken mutant, not a finding, so the mutant was fixed rather than the prediction reinterpreted.

Both cardinalities are stated up front so a MISSED cannot be reinterpreted afterwards as coverage.

    python scripts/mutation/listenerdiscovery.py
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from harness import harness

DISCOVERY = "scripts/listener-discovery.ts"
TEST = "tests/e2eTreeBinding.test.ts"

BINDING_TESTS = [TEST]
UNRELATED_TESTS = ["tests/recencyCardinality.test.ts"]

MUTATIONS = [
    (
        "M-DISCOVERY-AGREE the agreement rule always says the observations match",
        DISCOVERY,
        "  if (a === null || b === null) return false;\n"
        "  return a.pid === b.pid && a.identityToken === b.identityToken;",
        "  void a;\n  void b;\n  return true;",
    ),
    # M-DISCOVERY-FIRSTROW -- the defect review found on Windows, now a rule shared by both
    # platforms. Taking the first owner instead of requiring exactly one is how `Select-Object
    # -First 1` behaved: two distinct owners collapse to whichever the OS listed first, and row
    # order becomes authority. Expected: 2 red, the two-distinct-owner control and the
    # unparseable-row control, both of which exist for exactly this.
    (
        "M-DISCOVERY-FIRSTROW the first owner is taken instead of requiring exactly one",
        DISCOVERY,
        "  if (pids.some((p) => !Number.isInteger(p) || p <= 0)) return null;\n"
        "  const distinct = new Set(pids);\n"
        "  return distinct.size === 1 ? [...distinct][0] : null;",
        "  return pids[0];",
    ),
    (
        "M-DISCOVERY-RACE the second observation need not agree with the first",
        DISCOVERY,
        "  if (!observationsAgree(first, second)) return null;\n",
        "",
    ),
    (
        "M-DISCOVERY-SELF the current process is selected instead of the socket owner",
        DISCOVERY,
        '  const discover = process.platform === "win32" ? discoverWindows : discoverLinux;',
        "  const discover = (_p: number): ListenerIdentity | null => ({\n"
        "    pid: process.pid,\n"
        "    exe: null,\n"
        "    commandLine: null,\n"
        "    started: new Date(),\n"
        '    identityToken: "mutant",\n'
        '    authority: "mutant: current process, not the socket owner",\n'
        "  });",
    ),
]

sys.exit(harness([DISCOVERY], BINDING_TESTS, UNRELATED_TESTS, MUTATIONS, wall_seconds=900))

File: scripts/listener-discovery.ts
/**
 * WHICH PROCESS owns the socket listening on a loopback port — from OS evidence only.
 *
 * Split out of `e2e-tree-binding.ts` because it is the part with a platform matrix and the part
 * whose failure mode is silent. Everything here answers one question and refuses to answer it
 * loosely: a positive result must come from kernel-backed socket ownership, and anything short of
 * that returns null so the caller fails closed.
 *
 * ## WHY THIS EXISTS
 *
 * Discovery was Windows-only. CI runs on Linux, where it found nothing and every verdict collapsed
 * to UNPROVEN — so the check gated nothing in the environment that gates merges, and the tests
 * could only exercise injected decision logic. That gap was reported rather than papered over, and
 * this closes it.
 *
 * ## WHAT IS NOT EVIDENCE, and each of these was named as a way to get this wrong
 *
 * Process NAME or image. Command-line substrings. Working directory. Which `node` happens to be
 * running. Age. Whichever PID makes a test pass. A `node_modules` path shared between checkouts.
 * None of them identify a socket owner, and several would confidently select the wrong process on
 * this machine, where two worktrees share one `node_modules`.
 *
 * ## RACE AWARENESS
 *
 * A socket can be handed off and a PID can be reused between two observations. Identity is
 * therefore read TWICE around the lookup and must agree on both the PID and the process's own
 * start token; if it does not, the answer is null. A stable wrong answer is worse than no answer,
 * because the caller would treat it as evidence.
 */

import { execFileSync } from "node:child_process";
import { readFileSync, readdirSync, readlinkSync } from "node:fs";

export interface ListenerIdentity {
  pid: number;
  exe: string | null;
  commandLine: string | null;
  started: Date | null;
  /**
   * An exact, platform-native token for "this specific process instance".
   *
   * Compared across the two observations instead of the derived `started` Date, because the token
   * needs no arithmetic and therefore cannot drift: on Linux it is the raw `starttime` field in
   * clock ticks, on Windows the raw creation timestamp. A PID that was reused between observations
   * will not reproduce it.
   */
  identityToken: string;
  /** Which OS mechanism produced this, printed in the report so the reader can judge it. */
  authority: string;
}

function quiet(cmd: string, args: string[]): string | null {
  try {
    return execFileSync(cmd, args, {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
    }).trim();
  } catch {
    return null;
  }
}

/* ------------------------------------------------------------------ Windows */

function discoverWindows(port: number): ListenerIdentity | null {
  // The port is validated by the caller and interpolated as a bare integer; nothing from the URL
  // reaches the shell as text.
  const script =
    `$c = Get-NetTCPConnection -State Listen -LocalPort ${port} -ErrorAction SilentlyContinue | Select-Object -First 1; ` +
    `if (-not $c) { exit 0 }; ` +
    `$p = Get-CimInstance Win32_Process -Filter ("ProcessId=" + $c.OwningProcess) -ErrorAction SilentlyContinue; ` +
    `if (-not $p) { exit 0 }; ` +
    `[Console]::Out.Write(($p | Select-Object ProcessId,ExecutablePath,CommandLine,CreationDate | ConvertTo-Json -Compress))`;
  const out = quiet("powershell.exe", ["-NoProfile", "-NonInteractive", "-Command", script]);
  if (!out) return null;
  try {
    const j = JSON.parse(out) as {
      ProcessId: number;
      ExecutablePath: string | null;
      CommandLine: string | null;
      CreationDate: string | null;
    };
    if (typeof j.ProcessId !== "number") return null;
    // PowerShell renders CIM datetimes as `/Date(1234567890123)/` under ConvertTo-Json.
    const epoch = j.CreationDate ? /\/Date\((\d+)\)\//.exec(j.CreationDate)?.[1] : undefined;
    return {
      pid: j.ProcessId,
      exe: j.ExecutablePath ?? null,
      commandLine: j.CommandLine ?? null,
      started: epoch ? new Date(Number(epoch)) : j.CreationDate ? new Date(j.CreationDate) : null,
      identityToken: j.CreationDate ?? "",
      authority: "Get-NetTCPConnection -> Win32_Process (socket owner)",
    };
  } catch {
    return null;
  }
}

/* -------------------------------------------------------------------- Linux */

/**
 * The listening socket's inode, from `/proc/net/tcp` and `/proc/net/tcp6`.
 *
 * Columns: `sl local_address rem_address st ... inode`. `local_address` is `HEXIP:HEXPORT`, the
 * port big-endian. `st` is `0A` for LISTEN — checked, so an established connection to the same
 * port number is never mistaken for the listener.
 *
 * AMBIGUITY REFUSES. A port can be bound by more than one socket — separate v4 and v6 listeners,
 * or SO_REUSEPORT across processes. If the inodes resolve to more than one distinct owner the
 * answer is null, because "one of these two" is not an identification.
 */
function listeningInodes(port: number): string[] {
  const inodes: string[] = [];
  for (const file of ["/proc/net/tcp", "/proc/net/tcp6"]) {
    let text: string;
    try {
      text = readFileSync(file, "utf8");
    } catch {
      continue;
    }
    for (const line of text.split("\n").slice(1)) {
      const f = line.trim().split(/\s+/);
      if (f.length < 10) continue;
      const [, local, , state] = f;
      if (state !== "0A") continue;
      const hexPort = local.split(":")[1];
      if (hexPort === undefined || parseInt(hexPort, 16) !== port) continue;
      inodes.push(f[9]);
    }
  }
  return inodes;
}

/** The PIDs holding any of these socket inodes open, by walking `/proc/<pid>/fd`. */
function pidsHoldingInodes(inodes: Set<string>): Set<number> {
  const owners = new Set<number>();
  let entries: string[];
  try {
    entries = readdirSync("/proc");
  } catch {
    return owners;
  }
  for (const entry of entries) {
    if (!/^\d+$/.test(entry)) continue;
    const pid = Number(entry);
    let fds: string[];
    try {
      fds = readdirSync(`/proc/${pid}/fd`);
    } catch {
      // Another user's process, or one that exited mid-scan. Not readable is not a match, and it
      // is deliberately not an error either: the scan continues and a genuinely unreadable owner
      // ends as "no owner found", which fails closed.
      continue;
    }
    for (const fd of fds) {
      let target: string;
      try {
        target = readlinkSync(`/proc/${pid}/fd/${fd}`);
      } catch {
        continue;
      }
      const m = /^socket:\[(\d+)\]$/.exec(target);
      if (m && inodes.has(m[1])) {
        owners.add(pid);
        break;
      }
    }
  }
  return owners;
}

/** Boot time in epoch seconds, from `/proc/stat`. */
function bootTimeSeconds(): number | null {
  try {
    const m = /^btime\s+(\d+)$/m.exec(readFileSync("/proc/stat", "utf8"));
    return m ? Number(m[1]) : null;
  } catch {
    return null;
  }
}

/**
 * `starttime` (field 22 of `/proc/<pid>/stat`) as the raw identity token, plus an absolute Date.
 *
 * The comm field can contain spaces and parentheses, so everything before the LAST `)` is skipped
 * rather than split on whitespace from the left — a process named `foo bar) baz` would otherwise
 * shift every subsequent field.
 *
 * USER_HZ is 100. That is the kernel's userspace ABI constant for these fields, fixed independently
 * of CONFIG_HZ, so it is a documented interface rather than an assumption about this machine. The
 * derived Date is used only for start-order comparison; the RAW ticks are what identity is checked
 * against, so even if the derivation were wrong the race check would not be.
 */
function processStart(pid: number): { started: Date | null; token: string } | null {
  let stat: string;
  try {
    stat = readFileSync(`/proc/${pid}/stat`, "utf8");
  } catch {
    return null;
  }
  const close = stat.lastIndexOf(")");
  if (close < 0) return null;
  const fields = stat
    .slice(close + 2)
    .trim()
    .split(/\s+/);
  // After the comm field, `state` is index 0, so `starttime` (field 22 overall) is index 19.
  const ticks = fields[19];
  if (ticks === undefined || !/^\d+$/.test(ticks)) return null;
  const btime = bootTimeSeconds();
  const started = btime === null ? null : new Date((btime + Number(ticks) / 100) * 1000);
  return { started, token: ticks };
}

function readMaybe(path: string): string | null {
  try {
    return readFileSync(path, "utf8");
  } catch {
    return null;
  }
}

function discoverLinux(port: number): ListenerIdentity | null {
  const inodes = listeningInodes(port);
  if (inodes.length === 0) return null;

  const owners = pidsHoldingInodes(new Set(inodes));
  // Zero owners: the socket exists but no readable process holds it — another user, or it closed
  // mid-scan. More than one: genuinely ambiguous. Both fail closed.
  if (owners.size !== 1) return null;
  const pid = [...owners][0];

  const start = processStart(pid);
  if (start === null) return null;

  const exe = (() => {
    try {
      return readlinkSync(`/proc/${pid}/exe`);
    } catch {
      return null;
    }
  })();
  const cmdline = readMaybe(`/proc/${pid}/cmdline`);

  return {
    pid,
    exe,
    // `cmdline` is NUL-separated. Recorded for the report only; it is never evidence of identity.
    commandLine: cmdline === null ? null : cmdline.replace(/\0/g, " ").trim() || null,
    started: start.started,
    identityToken: start.token,
    authority: "/proc/net/tcp inode -> /proc/<pid>/fd (socket owner)",
  };
}

/* --------------------------------------------------------------- dispatcher */

/**
 * Discover the listener twice and require the two observations to agree.
 *
 * Returns null on: an unsupported platform, no listener, an unreadable or ambiguous owner, a parse
 * failure, or a PID/identity-token mismatch between the observations. The caller turns every one
 * of those into UNPROVEN, which is the only honest reading of "we could not tell".
 */
export function discoverListener(port: number): ListenerIdentity | null {
  if (!Number.isInteger(port) || port < 1 || port > 65535) return null;

  const discover = process.platform === "win32" ? discoverWindows : discoverLinux;
  if (process.platform !== "win32" && process.platform !== "linux") return null;

  const first = discover(port);
  if (first === null) return null;
  const second = discover(port);
  if (second === null) return null;
  if (first.pid !== second.pid || first.identityToken !== second.identityToken) return null;
  return first;
}
